src/main/java/com/uber/kafkaSpraynozzle/KafkaPoster.java

```diff
@@ -16,61 +23,102 @@
+        int leastIdx = 0;
+        for (int i = 1; i < responseTime.length; ++i){
+            if (responseTime[i] < responseTime[leastIdx]){
+                leastIdx = i;
+            }
+        }
+        return leastIdx;
     }
 
+    private int getBatchDataSize(List<ByteArrayEntity> batch){
+        int totalSize = 0;
+        for (int i = 0; i < batch.size(); ++i){
+            totalSize += (int)batch.get(i).getContentLength();
+        }
+        return totalSize;
+    }
+
+    private CloseableHttpClient client = null;
+    private long threadId = 0L;
+    private long lastReconnect = 0L;
+    private int[] responseTime = null;
+    private long[] responseTimestamp = null;
     public void run() {
-        long threadId = Thread.currentThread().getId();
+        threadId = Thread.currentThread().getId();
         System.out.println("Starting poster thread " + threadId);
-        CloseableHttpClient client = HttpClientBuilder.create().setConnectionManager(cm).build();
-        long lastReconnect = new Date().getTime();
+        client = HttpClientBuilder.create().setConnectionManager(cm).build();
+        lastReconnect = new Date().getTime();
+        responseTime = new int[urls.size()];
+        responseTimestamp = new long[urls.size()];
+
+        List<ByteArrayEntity> batch = new ArrayList<ByteArrayEntity>(batchSize + 1);
+
+        long totalPostingTimeMs = 0L;
+        long totalPostingCount = 0L;
+        long lastReportingTime = new Date().getTime();
         while(true) {
             ByteArrayEntity jsonEntity = queue.poll();
             if(jsonEntity != null) {
                 jsonEntity = messageFilter.filter(jsonEntity);
                 if (jsonEntity != null) {
-                    try {
-                        logQueue.add("posting");
-                        HttpPost post = new HttpPost(urls.get(currentUrl));
-                        currentUrl = (currentUrl + 1) % urls.size();
-                        post.setHeader("User-Agent", "KafkaSpraynozzle-0.0.1");
-                        post.setEntity(jsonEntity);
-                        CloseableHttpResponse response = client.execute(post);
-                        int statusCode = response.getStatusLine().getStatusCode();
-                        if (statusCode >= 200 && statusCode < 300) {
-                            logQueue.add("postSuccess");
-                        } else {
-                            logQueue.add("postFailure");
-                        }
-                        long currentTime = new Date().getTime();
-                        if(currentTime - lastReconnect > 10000) {
-                            lastReconnect = currentTime;
-                            response.close();
-                        } else {
-                            EntityUtils.consume(response.getEntity());
+                    batch.add(jsonEntity);
+
+                    if (batch.size() >= batchSize || getBatchDataSize(batch) >= batchMaxBytes) {
```

Comment:
Either we have accumulated enough items, or if the total data size is larger than (default=128KB) we can start to post.